referenceframe/model.go

```diff
@@ -321,47 +319,119 @@ func (m *SimpleModel) inputsToFrames(inputs []Input, collectAll bool) ([]*static
-		if pose == nil || (err != nil && !strings.Contains(err.Error(), OOBErrString)) {
+		if pose == nil || err != nil {
 			return nil, err
 		}
-		multierr.AppendInto(&err, errNew)
+
 		if collectAll {
 			var geometry spatialmath.Geometry
 			gf, err := transform.Geometries(input)
 			if err != nil {
 				return nil, err
 			}
-			geometries := gf.Geometries()
+
+			geometries := gf.geometries
 			if len(geometries) == 0 {
 				geometry = nil
 			} else {
 				geometry = geometries[0]
 			}
+
 			// TODO(pl): Part of the implementation for GetGeometries will require removing the single geometry restriction
 			fixedFrame, err := NewStaticFrameWithGeometry(transform.Name(), composedTransformation, geometry)
 			if err != nil {
 				return nil, err
 			}
+
 			poses = append(poses, fixedFrame.(*staticFrame))
 		}
+
 		composedTransformation = spatialmath.Compose(composedTransformation, pose)
 	}
+
 	// TODO(rb) as written this will return one too many frames, no need to return zeroth frame
 	poses = append(poses, &staticFrame{&baseFrame{"", []Limit{}}, composedTransformation, nil})
-	return poses, err
+	return poses, nil
+}
+
+// InputsToTransformOpt is like `inputsToFrames` but only returns the end effector pose. That allows
+// us to optimize away the recording of intermediate computations.
+func (m *SimpleModel) InputsToTransformOpt(inputs []Input) (spatialmath.Pose, error) {
+	if len(m.DoF()) != len(inputs) {
+		return nil, fmt.Errorf("wrong dimensions. DoF: %v Inputs: %v", len(m.DoF()), len(inputs))
+	}
+
+	// Start at ((1+0i+0j+0k)+(+0+0i+0j+0k)ϵ)
+	// composedTransformation := spatialmath.NewZeroPose()
+	composedTransformation := spatialmath.DualQuaternion{dualquat.Number{
+		Real: quat.Number{Real: 1},
+		Dual: quat.Number{},
+	}}
+	posIdx := 0
+
+	// get quaternions from the base outwards.
+	for _, transformI := range m.ordTransforms {
+		var pose spatialmath.Pose
+
+		switch transform := transformI.(type) {
+		case *staticFrame:
+			if len(transformI.DoF()) != 0 {
+				return nil, fmt.Errorf("unexpected DoF for staticFrame. DoF: %v", len(transformI.DoF()))
+			}
+
+			composedTransformation = spatialmath.DualQuaternion{
+				Number: composedTransformation.Transformation(transform.transform.(*spatialmath.DualQuaternion).Number),
+			}
+		case *rotationalFrame:
+			if len(transformI.DoF()) != 1 {
+				return nil, fmt.Errorf("unexpected DoF for rotationalFrame. DoF: %v", len(transformI.DoF()))
+			}
+
+			orientation := transform.InputToOrientation(inputs[posIdx])
+			pose = &spatialmath.DualQuaternion{
+				Number: dualquat.Number{
+					Real: orientation.Quaternion(),
+				},
+			}
+
+			posIdx++
+			composedTransformation = spatialmath.DualQuaternion{
+				Number: composedTransformation.Transformation(pose.(*spatialmath.DualQuaternion).Number),
+			}
+		default:
+			dof := len(transformI.DoF()) + posIdx
+			input := inputs[posIdx:dof]
+			posIdx = dof
+
+			var err error
+			pose, err = transform.Transform(input)
```

Comment:
`Transform()` converts the inputs into a pose in the frame of `transformI` relative to it's parent (the base of the arm) right?

Reply:
I believe so, yes